ncd-calculator/src/services/kgrid.ts

```diff
@@ -0,0 +1,328 @@
+import * as pako from 'pako';
+
+export interface GridObject {
+    id: string;
+    content: string;
+}
+
+export interface Position {
+    i: number;
+    j: number;
+}
+
+export interface Block {
+    topLeft: Position;
+    bottomRight: Position;
+}
+
+export interface GridState {
+    width: number;
+    height: number;
+    objects: GridObject[][];
+    ncdMatrix: number[][];
+    objectiveValue: number;
+}
+
+export const calculateNCD = (obj1: GridObject, obj2: GridObject): number => {
+    const str1 = obj1.content;
+
+    const str2 = obj2.content;
+
+    const data1 = new TextEncoder().encode(str1);
+    const data2 = new TextEncoder().encode(str2);
+
+    const combinedData = new Uint8Array(data1.length + data2.length);
+    combinedData.set(data1, 0);
+    combinedData.set(data2, data1.length);
+
+    const compressed1 = pako.deflate(data1);
+    const compressed2 = pako.deflate(data2);
+    const compressedCombined = pako.deflate(combinedData);
+
+    const C1 = compressed1.length;
+    const C2 = compressed2.length;
+    const C12 = compressedCombined.length;
+
+    const ncd = (C12 - Math.min(C1, C2)) / Math.max(C1, C2);
+
+    return ncd;
+}
+
+
+export const precomputeNCDMatrix = (objects: GridObject[]): number[][] => {
+    const n = objects.length;
+    const matrix: number[][] = Array(n).fill(0).map(() => Array(n).fill(0));
+
+    // Calculate NCD for each unique pair
+    for (let i = 0; i < n; i++) {
+        for (let j = i + 1; j < n; j++) {
+            const ncd = calculateNCD(objects[i], objects[j]);
+            // NCD is symmetric, so store it in both positions
+            matrix[i][j] = ncd;
+            matrix[j][i] = ncd;
+        }
+        // Distance to self is 0
+        matrix[i][i] = 0;
+    }
+
+    return matrix;
+}
+
+export const gradualFactor = (pos: Position): number => {
+    return Math.pow((1.01 * pos.i + 1.02 * pos.j + 10), 0.1);
+}
+
+export const calculateObjective = (grid: GridState): number => {
+    let total = 0;
+
+    // horizontal neighbors
+    for(let i = 0; i < grid.height; i++){
+        for(let j = 0; j < grid.width; j++) {
+            const right = (j + 1) % grid.width;
+
+            const obj1 = grid.objects[i][j];
+            const obj2 = grid.objects[i][right];
+
+            const ncdValue = grid.ncdMatrix[obj1.id][obj2.id];
+            const factor = gradualFactor({i, j});
+
+            total += ncdValue * factor;
```

Comment:
we only need to multiply by factor at the end because we add up the vertical and horizontal ncd for the same target. also these two loops can be combined